File: src/pft_FindPreviousBinaryMasks.m
function [ FormerBinaryMasksPresent, BinaryMaskStack, ReturnCode ] = pft_FindPreviousBinaryMasks(TopLevelOutputFolder, LeafFolder)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% A function to retrieve and re-use already created binary masks (in case the code for the FD calculation should change.             %
%                                                                                                                                    %
% The o/p files sought for are in fullfile(TopLevelOutputFolder, 'ThresImg', LeafFolder),                                           %
% and have names such as binaryMaskSlice1Phase1.png.                                                                                 %
%                                                                                                                                    %
% Outputs are:                                                                                                                       %
%                                                                                                                                    %
% FormerBinaryMasksPresent: a logical vector, one entry for each (possible) slice, numbered 1..20, but First and Last will be found. %
% BinaryMaskStack:          a 3-D logical image stack.                                                                               %
% ReturnCode:               a string with values { 'Folder for search does not exist.', ...                                          %
%                                                  'No former ROI''s found.', ...                                                    %
%                                                  'Binary mask stack not contiguous.' ...                                           %
%                                                  'OK' }.                                                                           %
% These are self-explanatory: (1) and (3) shouldn't happen in practice, (2) is possible and (4) means 'Success - go ahead.'.         %
%                                                                                                                                    %
% PFT - 01. 12. 2016.                                                                                                                %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% A global constant which matches the auto-FD code, but any number large enough in practice will suffice
NSLICES = 20;

% Default o/p for immediate return
FormerBinaryMasksPresent = false([1, NSLICES]);
BinaryMaskStack = [];
ReturnCode = 'Folder for search does not exist.';

% Check that the folder to be searched actually exists
if (exist(fullfile(TopLevelOutputFolder, 'ThresImg', LeafFolder), 'dir') ~= 7)  
  return;
end

% Check for the presence of end-diastole files only
for n = 1:NSLICES
  FileName = sprintf('binaryMaskSlice%1dPhase1.png', n);
  if (exist(fullfile(TopLevelOutputFolder, 'ThresImg', LeafFolder, FileName), 'file') == 2)
    FormerBinaryMasksPresent(n) = true;
  end
end

% Return if no previous ROI's are found
if (sum(FormerBinaryMasksPresent) == 0)
  ReturnCode = 'No former ROI''s found.';
  return;
end  

% Now check for contiguity
First = find(FormerBinaryMasksPresent == true, 1, 'first');
Last  = find(FormerBinaryMasksPresent == true, 1, 'last');

M = Last - First + 1;
N = sum(FormerBinaryMasksPresent(First:Last));

if (M ~= N)
  ReturnCode = 'Binary mask stack not contiguous.';
  return;
end

% Finally, extract the stack of binary masks - use the 'first' one present to determine the image dimensions
FileName = sprintf('binaryMaskSlice%1dPhase1.png', First);
PathName = fullfile(TopLevelOutputFolder, 'ThresImg', LeafFolder, FileName);
BM = logical(imread(PathName));

[ NR, NC ] = size(BM);

BinaryMaskStack = false([NR, NC, NSLICES]);

for p = First:Last
  if (p == First)
    BinaryMaskStack(:, :, p) = BM;
  else
    FileName = sprintf('binaryMaskSlice%1dPhase1.png', p);
    PathName = fullfile(TopLevelOutputFolder, 'ThresImg', LeafFolder, FileName);
    BM = logical(imread(PathName));
    BinaryMaskStack(:, :, p) = BM;
  end
end

ReturnCode = 'OK';
    
end

File: python/src/pft_FindPreviousBinaryMasks.py
import os
import numpy as np
from skimage import io

def pft_find_previous_binary_masks(top_level_output_folder, leaf_folder):
    nslices = 20
    former_binary_masks_present = [False] * nslices
    binary_mask_stack = []
    return_code = "Folder for search does not exist."

    thres_img_folder = os.path.join(top_level_output_folder, 'ThresImg', leaf_folder)
    if not os.path.isdir(thres_img_folder):
        return former_binary_masks_present, binary_mask_stack, return_code

    for n in range(nslices):
        file_name = f'binaryMaskSlice{n+1}Phase1.png'
        if os.path.isfile(os.path.join(thres_img_folder, file_name)):
            former_binary_masks_present[n] = True

    if not any(former_binary_masks_present):
        return_code = "No former ROI's found."
        return former_binary_masks_present, binary_mask_stack, return_code

    first = next(i for i, x in enumerate(former_binary_masks_present) if x)
    last = len(former_binary_masks_present) - next(i for i, x in enumerate(reversed(former_binary_masks_present)) if x) - 1

    m = last - first + 1
    n = sum(former_binary_masks_present[first:last+1])
    
    if m != n:
        return_code = 'Binary mask stack not contiguous.'
        return former_binary_masks_present, binary_mask_stack, return_code

    binary_mask_stack = np.zeros((nslices, nslices, nslices), dtype=bool)
    for p in range(first, last+1):
        file_name = f'binaryMaskSlice{p+1}Phase1.png'
        path_name = os.path.join(thres_img_folder, file_name)
        bm = io.imread(path_name).astype(bool)
        binary_mask_stack[:, :, p] = bm

    return_code = 'OK'
    return former_binary_masks_present, binary_mask_stack, return_code
